### PR TITLE
refactor(android): Move build-publish.sh to builder script

### PR DESCRIPTION
**Pull request by darcywong00**
In preparation for #12325

This refactors the `build-publish.sh` script that publishes the Keyman and FirstVoices apk's to the Play Store into the corresponding builder scripts.

I noticed the top level android build.sh script already declares a publish action
https://github.com/keymanapp/keyman/blob/fdad0ca2bb3932a6389738882554042b4802672e/android/build.sh#L21

But it must be a NOOP.

Will need to update CI configs and :cherries: pick to stable-17.0

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by darcywong00** [comment]
> Will need to update CI configs and 🍒 pick to stable-17.0

Hmm, stable-17.0 doesn't have the refactored builder scripts from #11329
I'm more inclined to keep the TC 17.0 publish step intact - though both branches have .build-builder file...

File: android/KMAPro/build.sh [hunk 1]
```diff
@@ -109,3 +113,15 @@ if builder_start_action test; then
...
   builder_finish_action success test
 fi
+
+if builder_start_action publish; then
+  # Copy Release Notes
+  generateReleaseNotes
+
+  # Publish Keyman for Android to Play Store
+  echo "PUBLISH_FLAGS $PUBLISH_FLAGS"
+  cd "$KEYMAN_ROOT/android/KMAPro/"
```
Comment (mcdurdin) [inline android/KMAPro/build.sh:123]: This shouldn't be necessary as the script already `cd $THIS_SCRIPT_PATH` when it starts (via builder)

```suggestion
```

File: android/KMAPro/build.sh [hunk 2]
```diff
@@ -109,3 +113,15 @@ if builder_start_action test; then
...
+  ./gradlew $DAEMON_FLAG $PUBLISH_FLAGS
```
Comment (mcdurdin) [inline android/KMAPro/build.sh:124]: I wonder if `$PUBLISH_FLAGS` is necessary if it never changes? Just put it inline in the script and reduce the set of variables floating about.

```suggestion
  ./gradlew $DAEMON_FLAG publishReleaseApk
```

File: android/KMAPro/build.sh [hunk 3]
```diff
@@ -109,3 +113,15 @@ if builder_start_action test; then
 
   builder_finish_action success test
 fi
+
+if builder_start_action publish; then
+  # Copy Release Notes
+  generateReleaseNotes
+
+  # Publish Keyman for Android to Play Store
+  echo "PUBLISH_FLAGS $PUBLISH_FLAGS"
```
Comment (mcdurdin) [inline android/KMAPro/build.sh:122]:
```suggestion
```

File: android/KMAPro/build.sh [hunk 4]
```diff
@@ -10,13 +10,16 @@ THIS_SCRIPT="$(readlink -f "${BASH_SOURCE[0]}")"
...
+. "$KEYMAN_ROOT/android/KMAPro/build-play-store-notes.inc.sh"
+
 # ################################ Main script ################################
 
 # Definition of global compile constants
 
 CONFIG="release"
 BUILD_FLAGS="build -x lint -x test"                     # Gradle build w/o test
 TEST_FLAGS="-x assembleRelease lintRelease testRelease" # Gradle test w/o build
+PUBLISH_FLAGS="publishReleaseApk"                       # Gradle publish
```
Comment (mcdurdin) [inline android/KMAPro/build.sh:22]:
```suggestion
```

File: oem/firstvoices/android/build.sh [hunk 5]
```diff
@@ -85,3 +87,12 @@ if builder_start_action test; then
...
 fi
+
+if builder_start_action publish; then
+  echo "PUBLISH_FLAGS $PUBLISH_FLAGS"
+
+  cd "$KEYMAN_ROOT/oem/firstvoices/android/"
+  ./gradlew $DAEMON_FLAG $PUBLISH_FLAGS  
+
+  builder_finish_action success publish
+fi
```
Comment (mcdurdin) [inline oem/firstvoices/android/build.sh:98]:
```suggestion
builder_run_action publish    ./gradlew $DAEMON_FLAG publishReleaseApk
```

File: oem/firstvoices/android/build.sh [hunk 6]
```diff
@@ -17,6 +17,7 @@ THIS_SCRIPT="$(readlink -f "${BASH_SOURCE[0]}")"
 CONFIG="release"
 BUILD_FLAGS="build -x lint -x test"                     # Gradle build w/o test
 TEST_FLAGS="-x assembleRelease lintRelease testRelease" # Gradle test w/o build
+PUBLISH_FLAGS="publishReleaseApk"                       # Gradle publish
```
Comment (mcdurdin) [inline oem/firstvoices/android/build.sh:20]:
```suggestion
```

**Review by mcdurdin — APPROVED** [review]
LGTM, just simplified the changes to remove unnecessary variables -- if the variable never changes, then it's questionable whether it should be a variable (and consts don't buy us anything here either)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.108-alpha](https://keyman.com/downloads/releases/alpha/18.0.108)

**Pull request merged**